File: 0189_rotate_array.go
func reverse(nums []int, start int, end int) {
    for start < end {
        temp := nums[start]
        nums[start] = nums[end]
        nums[end] = temp
        start++
        end--
    }
}

func rotate(nums []int, k int)  {
    n := len(nums)
    k = k % n
    if k == 0 {
        return
    }

    reverse(nums, 0, n - 1)
    reverse(nums, 0, k - 1)
    reverse(nums, k, n - 1)
}

File: 0189_rotate_array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    k = k % nums.length;
    if (k === 0) return;

    reverse(nums, 0, nums.length - 1);
    reverse(nums, 0, k - 1);
    reverse(nums, k, nums.length - 1)
};

function reverse(array, start, end) {
    while(start < end) {
        let temp = array[start];
        array[start] = array[end];
        array[end] = temp;
        start++;
        end--;
    }
}

